ddl/ddl.go

```diff
@@ -1425,85 +1425,229 @@ func get2JobsFromTable(sess *sess.Session) (*model.Job, *model.Job, error) {
+
+		// There may be some conflict during the update, try it again
+		if ns.Commit() != nil {
 			continue
 		}
-		if !job.IsRollbackable() {
-			errs[i] = dbterror.ErrCannotCancelDDLJob.GenWithStackByArgs(job.ID)
-			continue
+
+		for id, idx := range jobMap {
+			errs[idx] = dbterror.ErrDDLJobNotFound.GenWithStackByArgs(id)
 		}
-		job.State = model.JobStateCancelling
-		// Make sure RawArgs isn't overwritten.
-		err := json.Unmarshal(job.RawArgs, &job.Args)
+
+		break
+	}
+	return errs, nil
+}
+
+// CancelJobs cancels the DDL jobs according to user command.
+func CancelJobs(se sessionctx.Context, ids []int64) (errs []error, err error) {
+	return processJobs(cancelRunningJob, se, ids, model.AdminCommandByEndUser)
+}
+
+// PauseJobs pause all the DDL jobs according to user command.
+func PauseJobs(se sessionctx.Context, ids []int64) ([]error, error) {
+	return processJobs(pauseRunningJob, se, ids, model.AdminCommandByEndUser)
+}
+
+// ResumeJobs resume all the DDL jobs according to user command.
+func ResumeJobs(se sessionctx.Context, ids []int64) ([]error, error) {
+	return processJobs(resumePausedJob, se, ids, model.AdminCommandByEndUser)
+}
+
+// CancelJobsBySystem , internal only
+func CancelJobsBySystem(se sessionctx.Context, ids []int64) (errs []error, err error) {
+	return processJobs(cancelRunningJob, se, ids, model.AdminCommandBySystem)
+}
+
+// PauseJobsBySystem , internal only
+func PauseJobsBySystem(se sessionctx.Context, ids []int64) (errs []error, err error) {
+	return processJobs(pauseRunningJob, se, ids, model.AdminCommandBySystem)
+}
+
+// ResumeJobsBySystem , internal only
+func ResumeJobsBySystem(se sessionctx.Context, ids []int64) (errs []error, err error) {
+	return processJobs(resumePausedJob, se, ids, model.AdminCommandBySystem)
+}
+
+// processAllJobs , process all Jobs from the job table, 100 jobs each time in case of high memory usage
+func processAllJobs(process func(*sess.Session, *model.Job, model.AdminCommandOperator) (err error),
+	se sessionctx.Context, byWho model.AdminCommandOperator) (map[int64]error, error) {
+	var err error
+	var jobErrs map[int64]error
+
+	ns := sess.NewSession(se)
+	err = ns.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	var jobID int64 = 0
+	var limit int64 = 100
+	for {
+		jobs, err := getJobsBySQL(ns, JobTable,
+			fmt.Sprintf("job_id >= %s order by job_id limit %s",
+				strconv.FormatInt(jobID, 10),
+				strconv.FormatInt(limit, 10)))
 		if err != nil {
-			errs[i] = errors.Trace(err)
-			continue
+			ns.Rollback()
+			return nil, err
 		}
-		err = updateDDLJob2Table(sessCtx, job, true)
-		if err != nil {
-			errs[i] = errors.Trace(err)
+
+		for _, job := range jobs {
+			err := process(ns, job, byWho)
+			if err != nil {
+				ns.Rollback()
+				return jobErrs, err
+			}
 		}
+
+		// If rows returned is smaller than $limit, then there is no more records
+		if len(jobs) < 100 {
+			break
+		}
+
+		jobID += limit
```

Comment:
JobID is not linear increament, it is shared used by table id and index id assign.